### ISSUE TITLE
Appium 1.8+ is brutally slow using Geb and Spock framework on iOS because AppiumDriver.getSession() became slow

### ISSUE THREAD

**Issue by jacattau** [appium/appium#10790]
## The problem

It appears something changed between Appium 1.7.2 and 1.8.0 in how Appium interacts with Geb/Spock framework.  On Appium 1.6.5 and 1.7.2 I had no issues automating a hybrid app on iOS and the speed was pretty stellar.  However, after updating to 1.8.0, 1.8.1, and 1.8.2-beta, the performance has absolutely tanked.  The same tests on 1.6.5 and 1.7.2 that took 5 seconds now takes upwards of a minute since 1.8.0.  Android is unaffected, the speed with Geb/Spock is still great.  This is only affecting iOS and the mobile safari browser.

NOTE:  Selenium performance is normal.

## Environment

* Appium version (or git revision) that exhibits the issue: 1.8.0 and above.  I did not test beta versions between 1.7.2 and 1.8.0
* Last Appium version that did not exhibit the issue (if applicable): 1.6.5, 1.7.2
* Desktop OS/version used to run Appium: MacOS High Sierra
* Node.js version (unless using Appium.app|exe): 9.6.1
* Mobile platform/version under test: iOS 11.3.1
* Real device or emulator/simulator: Real Device
* Appium CLI or Appium.app|exe: CLI

## Details

I have an example test written with Geb/Spock:

```
def "should go to Individual button"(){
        given:
        page = waitFor { at HomePage }

        when:
        $(id: "cwl").click()
        waitFor { $(id: "state") }
        enterInput($(id: "age"),"18")

        then:
        waitFor { $(id: "ageError").displayed }
        $(id: "ageError").text() == "Please enter an age between 0 and 17."
    }
```

This test takes about 3 seconds on a desktop browser and about 5 seconds on a mobile device using Appium 1.6.5 and Appium 1.7.2.  This very same test on Appium 1.8+ on iOS takes about 1 minute.  Android is still 5 seconds.

## Link to Appium logs
[Log using Appium 1.7.2 working as expected](https://gist.github.com/jacattau/c3d953b58868a6d4478ef9bebaf3b1a4)
[Log using Appium 1.8.2-beta, very slow](https://gist.github.com/jacattau/a74a874c83d9d50fa8c1206cc5946bda)

## Code To Reproduce Issue [ Good To Have ]

This is internal to our company app so not sure how much help this is but:
```
def "should go to Individual button"(){
        given:
        page = waitFor { at HomePage }

        when:
        $(id: "cwl").click()
        waitFor { $(id: "state") }
        enterInput($(id: "age"),"18")

        then:
        waitFor { $(id: "ageError").displayed }
        $(id: "ageError").text() == "Please enter an age between 0 and 17."
    }
```

**Comment by mykola-mokhnach** [comment]
Have you tried to investigate the problem yourself? What exact calls take longer than before? Is implicit wait wait set to a proper value?

**Comment by jacattau** [comment]
I have tried investigating myself and opening this issue has become my last resort.  The waitFor's are set to 5 seconds.  Going through debugging through IntelliJ and using println after each step, it just appears that every call is taking much longer than before.  I've dug as far as I can understand the code and event taking it step by step and stepping down through the code there are just calls that hang for a few seconds before proceeding.

Given the fact that there are no issues with the 2 release versions of appium and it started becoming an issue recently I figured there had to be something going on behind the scenes with Appium.

**Comment by jacattau** [comment]
[timestamp logs on 1.8.2-beta](https://gist.github.com/jacattau/5ac31595dc8d0df6ba16c7887dc441b9)

I am the same test twice.  Looks like its ~60 seconds to launch and run the first test, and then the 2nd (same test) takes about 30 seconds.  5 seconds to 30 seconds is quite a performance hit, especially when you span it for 600+ tests.

**Comment by KazuCocoa** [comment]
Does this happen only Real Device?
(Doesn't happen Simulator?)

Since it's easy to run tests with a simulator. So if it happens in a simulator as well, I'd like to ensure this with it.

**Comment by KazuCocoa** [comment]
I see. So far, I also haven't heard such issue by users. But I'll try to set test environment up and run tests a couple of times.

If the issue happens with `newWDA: true` and Appium 1.7.x, we probably can assume WDA affects sometimes.

**Comment by jacattau** [comment]
@KazuCocoa I can confirm I experience the same slowness on a simulator.

**Comment by KazuCocoa** [comment]
Compared a number of `Webkit response timeout: 5000`, you can see it in `1.8.0` three times than `1.7.2`.

In my locale, I ran `ruby_lib_core`'s test suite and compare the running time. (Almost them are native app) But the running time was almost same.

Thus, the cause is `RemoteDebugger` related.

**Comment by KazuCocoa** [comment]
oh, you already commented on https://github.com/google/ios-webkit-debug-proxy/issues/258#issuecomment-391419500 about the webkit debugger.

**Comment by jacattau** [comment]
Yes I have the latest webkit from that issue.  With my setup completely the same as it is, I can literally switch between 1.7.2 and 1.8.2-beta and see the difference in the performance.

**Comment by KazuCocoa** [comment]
I have no good insight for the proxy...

@mykola-mokhnach @imurchie Do you have any idea in xcuitest driver associated with the proxy?
(Or we didn't change them?)

**Comment by jacattau** [comment]
A colleague of mine noticed these differences in the appium logs when troubleshooting with me:

With 1.8.2-beta
```[Appium] Creating new XCUITestDriver (v2.84.0) session```

and with 1.7.2
```[Appium] Creating new XCUITestDriver (v2.64.0) session```

So possibility of that version change in the XCUITestDriver?

Another thing he noticed the logging lines triples and this line is different in the beta:
```[debug] [RemoteDebugger] Received result for atom 'get_window_size' execution: {"width":375,"height":647}```

Not sure if any of this is helpful, but wanted to get some discussion going.

**Comment by KazuCocoa** [comment]
What about `useNewWDA: true` capability with Appium 1.7.2?

> If the issue happens with `useNewWDA: true` and Appium 1.7.x, we probably can assume WDA affects sometimes.

The capability tries to use the latest WDA. The result helps to split the issue is in WDA or out of WDA.

**Comment by jacattau** [comment]
@KazuCocoa just did a comparison all with the same 2 tests.

Appium 1.7.2 **WITH** useNewWDA capability took 57s 787ms.
Appium 1.7.2 _WITHOUT_ useNewWDA capability took 46s 25ms.
Appium 1.8.2-beta **WITH** useNewWDA capability took 1m 37s 990ms.
Appium 1.8.2-beta _WITHOUT_ useNewWDA capability took 1m 40s 647ms.

**Comment by KazuCocoa** [comment]
thanks!
I found the issue in the WDA repository. https://github.com/facebook/WebDriverAgent/issues/780

What about the time when you call _pageSource_?
If the times are different (and slow in 1.8.2), the getting elements has an issue.

---

📝 

https://github.com/appium/appium-xcuitest-driver/blob/af84eaa70bc48745229a04bbd2c8b16505fc34aa/lib/commands/general.js#L103 is probably `[debug] [RemoteDebugger] Received result for atom 'get_window_size'` and the code exists over 2 years..

**Comment by jacattau** [comment]
Are you asking if there is a time difference in calling `driver.getPageSource()` between 1.7.2 and 1.8.2-beta?  I just did a test for both versions calling it and printing it out and there wasn't any differenceas both were near instant.  It's when interacting with the elements that 1.8+ is very slow.

**Comment by KazuCocoa** [comment]
Yes, and thanks. If the test page is complex and the time is difference, I thought the time difference happened.
hmmm...

I tried to call the below methods at [here](https://github.com/appium/ruby_lib/blob/cb9450d207fb2e23b4dfef4318f47ea3d04c7689/ios_tests/lib/ios/specs/common/web_context.rb#L19) and measure the running time rouphly, and the running time was alomost same.

```
    now = Time.now
    (1..10).each { |_v| id("gbw").displayed }
    puts Time.now - now
```

The test page is Apple's site.

![image](https://user-images.githubusercontent.com/5511591/41047011-8b36674c-69e6-11e8-8e71-c5c236428fbd.png)

I sometimes got different pagesource even they were same webpage. 🤔

**Comment by jacattau** [comment]
Any update on this?  Are you waiting on something from me?

**Comment by KazuCocoa** [comment]
Not special.

What about disabling https://github.com/appium/appium-xcuitest-driver/blob/3aa990ed135a1397d051e383e5edc849e9adf33f/lib/desired-caps.js#L136 ?

We can find safari console logs.
```
[debug] [RemoteDebugger] WebKit debugger web socket connected to url: ws://localhost:27753/devtools/page/1
[debug] [RemoteDebugger] Starting to listen for JavaScript console
[debug] [RemoteDebugger] Sending WebKit data: {"method":"Console.enable","params":{"objectGroup":"console","includeCommandLineAPI":true,"doNotPauseOnExceptionsAndMuteConsole":true}}
[debug] [RemoteDebugger] Webkit response timeout: 5000
[debug] [RemoteDebugger] Receiving WebKit data: '{"result":{},"id":1}'
[debug] [RemoteDebugger] Found handler for message '1'
```

**Comment by jacattau** [comment]
Are you asking me to run tests with that showSafariConsoleLog as false?  I'm not sure how to do that as I didn't see a capability for it.

**Comment by KazuCocoa** [comment]
I tried to disable `Starting to listen for JavaScript console` on my local via safari browser, but I couldn't only capabilities, sorry.

**Comment by KazuCocoa** [comment]
Maybe I figured out. Performance of `@driver.session_capabilities`, `AppiumDriver.getSession()`,  became slow in 1.8.0. I'm not familiar with Java client, but Java client calls `AppiumDriver.getSession()` many times. So, the performance slow so much.

---

## 1.7.2

```
[HTTP] --> GET /wd/hub/session/adbc5295-8238-46dc-b091-83b6bb965df2 {}
[debug] [MJSONWP] Calling AppiumDriver.getSession() with args: ["adbc5295-8238-46dc-b091-83b6bb965df2"]
[debug] [XCUITest] Executing command 'getSession'
[debug] [JSONWP Proxy] Proxying [GET /] to [GET http://localhost:8100/session/7557E52D-2871-49FD-B99F-27E687020368] with no body
[debug] [JSONWP Proxy] Got response with status 200: "{\n  \"value\" : {\n    \"sessionId\" : \"7557E52D-2871-49FD-B99F-27E687020368\",\n    \"capabilities\" : {\n      \"device\" : \"iphone\",\n      \"browserName\" : \"UICatalog\",\n      \"sdkVersion\" : \"10.3.1\",\n      \"CFBundleIdentifier\" : \"com.example.apple-samplecode.UICatalog\"\n    }\n  },\n  \"sessionId\" : \"7557E52D-2871-49FD-B99F-27E687020368\",\n  \"status\" : 0\n}"
[XCUITest] Merging WDA caps over Appium caps for session detail response
[debug] [MJSONWP] Responding to client with driver.getSession() result: {"udid":"DED4DBAD-8E5E-4AD6-BDC4-E75CF9AD84D8","platformName":"ios","automationName":"XCUITest","app":"/Users/kazuaki/GitHub/ruby_lib_core/test/functional/app/UICatalog.app","platformVersion":"10.3","deviceName":"iPhone Simulator","useNewWDA":true,"useJSONSource":true,"someCapability":"some_capability","newCommandTimeout":120,"device":"iphone","browserName":"UICatalog","sdkVersion":"10.3.1","CFBundleIdentifier":"com.example.apple-samplecode.UICatalog"}
```

## 1.8.x

```
[HTTP] --> GET /wd/hub/session/422f60d3-0f73-4bb8-9467-edd65be05d14
[HTTP] {}
[debug] [MJSONWP] Calling AppiumDriver.getSession() with args: ["422f60d3-0f73-4bb8-9467-edd65be05d14"]
[debug] [XCUITest] Executing command 'getSession'
[debug] [JSONWP Proxy] Proxying [GET /] to [GET http://localhost:8100/session/12234C84-DF1C-4BBC-8ED8-278F67681106] with no body
[debug] [JSONWP Proxy] Got response with status 200: "{\n  \"value\" : {\n    \"sessionId\" : \"12234C84-DF1C-4BBC-8ED8-278F67681106\",\n    \"capabilities\" : {\n      \"device\" : \"iphone\",\n      \"browserName\" : \"UICatalog\",\n      \"sdkVersion\" : \"10.3.1\",\n      \"CFBundleIdentifier\" : \"com.example.apple-samplecode.UICatalog\"\n    }\n  },\n  \"sessionId\" : \"12234C84-DF1C-4BBC-8ED8-278F67681106\",\n  \"status\" : 0\n}"
[XCUITest] Merging WDA caps over Appium caps for session detail response
[debug] [JSONWP Proxy] Proxying [GET /wda/screen] to [GET http://localhost:8100/session/12234C84-DF1C-4BBC-8ED8-278F67681106/wda/screen] with no body
[debug] [JSONWP Proxy] Got response with status 200: "{\n  \"value\" : {\n    \"statusBarSize\" : {\n      \"width\" : 375,\n      \"height\" : 23.4375\n    },\n    \"scale\" : 2\n  },\n  \"sessionId\" : \"12234C84-DF1C-4BBC-8ED8-278F67681106\",\n  \"status\" : 0\n}"
[debug] [JSONWP Proxy] Proxying [GET /wda/screen] to [GET http://localhost:8100/session/12234C84-DF1C-4BBC-8ED8-278F67681106/wda/screen] with no body
[debug] [JSONWP Proxy] Got response with status 200: "{\n  \"value\" : {\n    \"statusBarSize\" : {\n      \"width\" : 375,\n      \"height\" : 23.4375\n    },\n    \"scale\" : 2\n  },\n  \"sessionId\" : \"12234C84-DF1C-4BBC-8ED8-278F67681106\",\n  \"status\" : 0\n}"
[debug] [JSONWP Proxy] Proxying [GET /wda/screen] to [GET http://localhost:8100/session/12234C84-DF1C-4BBC-8ED8-278F67681106/wda/screen] with no body
[debug] [JSONWP Proxy] Got response with status 200: "{\n  \"value\" : {\n    \"statusBarSize\" : {\n      \"width\" : 375,\n      \"height\" : 23.4375\n    },\n    \"scale\" : 2\n  },\n  \"sessionId\" : \"12234C84-DF1C-4BBC-8ED8-278F67681106\",\n  \"status\" : 0\n}"
[debug] [JSONWP Proxy] Proxying [GET /window/size] to [GET http://localhost:8100/session/12234C84-DF1C-4BBC-8ED8-278F67681106/window/size] with no body
[debug] [JSONWP Proxy] Got response with status 200: "{\n  \"value\" : {\n    \"width\" : 375,\n    \"height\" : 667\n  },\n  \"sessionId\" : \"12234C84-DF1C-4BBC-8ED8-278F67681106\",\n  \"status\" : 0\n}"
[debug] [MJSONWP] Responding to client with driver.getSession() result: {"udid":"DED4DBAD-8E5E-4AD6-BDC4-E75CF9AD84D8","platformName":"ios","automationName":"XCUITest","app":"/Users/kazuaki/GitHub/ruby_lib_core/test/functional/app/UICatalog.app","platformVersion":"10.3","deviceName":"iPhone Simulator","useNewWDA":true,"useJSONSource":true,"someCapability":"some_capability","newCommandTimeout":120,"device":"iphone","browserName":"UICatalog","sdkVersion":"10.3.1","CFBundleIdentifier":"com.example.apple-samplecode.UICatalog","pixelRatio":2,"statBarHeight":23.4375,"viewportRect":{"left":0,"top":47,"width":750,"height":1287}}
[HTTP] <-- GET /wd/hub/session/422f60d3-0f73-4bb8-9467-edd65be05d14 200 1221 ms - 628
[HTTP]
```

**Comment by mykola-mokhnach** [comment]
I'll take care about memoizing the result of `wda/screen` call

**Comment by KazuCocoa** [comment]
I think https://github.com/appium/java-client/pull/769 is the fix to reduce a number of calling `getSession` in Java client.

@jacattau What version of Java client do you use? If you use the latest Java client, the performance will improve.

**Comment by jacattau** [comment]
I'm currently using 1.8.0_152.  I will update and retest.

**Comment by jacattau** [comment]
I will update now and test.  We are using 5.0.4 currently.

**Comment by jacattau** [comment]
Tried 6.0.0 and iOS is much, much faster!

**Comment by mykola-mokhnach** [comment]
@jacattau Can you also try to reinstall appium@beta (appium-remote-debugger should be at version 3.13.0+) and try your tests? We've resolved some timeout issues, so some stuff should perform faster now on server side.

**Comment by jacattau** [comment]
@mykola-mokhnach Yes, it appears it is faster.  This is great work.

**Comment by dalong417** [comment]
Is this included in 1.9.0?

**Comment by KazuCocoa** [comment]
Yup. 1.9.0 and the latest Java client have resolved this issue
